mintscript/src/translate.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::chain::{
+    block::BlockRewardTransactable, signature::inputsig::InputWitness, tokens::TokenId,
+    AccountCommand, AccountOutPoint, AccountSpending, DelegationId, Destination, PoolId,
+    SignedTransaction, TxOutput, UtxoOutPoint,
+};
+use pos_accounting::PoSAccountingView;
+use tokens_accounting::TokensAccountingView;
+use utxo::Utxo;
+
+use crate::WitnessScript;
+
+/// An error that can happen during translation of an input to a script
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum TranslationError {
+    #[error("Attempt to spend an unspendable output")]
+    Unspendable,
+
+    #[error("Illegal account spend")]
+    IllegalAccountSpend,
+
+    #[error("Illegal output spend")]
+    IllegalOutputSpend,
+
+    #[error(transparent)]
+    PoSAccounting(#[from] pos_accounting::Error),
+
+    #[error(transparent)]
+    TokensAccounting(#[from] tokens_accounting::Error),
+
+    #[error("Stake pool {0} does not exist")]
+    PoolNotFound(PoolId),
+
+    #[error("Delegation {0} does not exist")]
+    DelegationNotFound(DelegationId),
+
+    #[error("Token with id {0} does not exist")]
+    TokenNotFound(TokenId),
+}
+
+/// Contextual information about given input
+pub enum InputInfo<'a> {
+    Utxo {
+        outpoint: &'a UtxoOutPoint,
+        utxo: Utxo,
+    },
+    Account {
+        outpoint: &'a AccountOutPoint,
+    },
+    AccountCommand {
+        command: &'a AccountCommand,
+    },
+}
+
+impl InputInfo<'_> {
+    pub fn as_utxo_output(&self) -> Option<&TxOutput> {
+        match self {
+            InputInfo::Utxo { outpoint: _, utxo } => Some(utxo.output()),
+            InputInfo::Account { .. } | InputInfo::AccountCommand { .. } => None,
+        }
+    }
+}
+
+pub trait InputInfoProvider {
+    fn input_info(&self) -> &InputInfo;
+    fn witness(&self) -> &InputWitness;
+}
+
+/// Provides information necessary to translate an input to a script.
+pub trait SignatureInfoProvider: InputInfoProvider {
+    type PoSAccounting: PoSAccountingView<Error = pos_accounting::Error>;
+    type Tokens: TokensAccountingView<Error = tokens_accounting::Error>;
+
+    fn pos_accounting(&self) -> &Self::PoSAccounting;
+    fn tokens(&self) -> &Self::Tokens;
```

Comment:
Can't we add UtxosView here?

Reply:
So we had a discussion about this yesterday, and it seems that there's some abstraction leakage and some broken symmetry. But the problem is that UtxosView is needed earlier to be able to calculate signatures.